solidity/contracts/isms/hook/AbstractMessageIdAuthorizedIsm.sol

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+/*@@@@@@@       @@@@@@@@@
+ @@@@@@@@@       @@@@@@@@@
+  @@@@@@@@@       @@@@@@@@@
+   @@@@@@@@@       @@@@@@@@@
+    @@@@@@@@@@@@@@@@@@@@@@@@@
+     @@@@@  HYPERLANE  @@@@@@@
+    @@@@@@@@@@@@@@@@@@@@@@@@@
+   @@@@@@@@@       @@@@@@@@@
+  @@@@@@@@@       @@@@@@@@@
+ @@@@@@@@@       @@@@@@@@@
+@@@@@@@@@       @@@@@@@@*/
+
+// ============ Internal Imports ============
+
+import {IInterchainSecurityModule} from "../../interfaces/IInterchainSecurityModule.sol";
+import {LibBit} from "../../libs/LibBit.sol";
+import {Message} from "../../libs/Message.sol";
+import {TypeCasts} from "../../libs/TypeCasts.sol";
+
+// ============ External Imports ============
+
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+/**
+ * @title ArbtractNativeISM
+ * @notice Uses the native bridges to verify interchain messages.
+ * @dev In the future, the hook might be moved inside the Mailbox which doesn't require storage mappings for senders.
+ *      for more details see https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2381
+ * @dev V3 WIP
+ */
+abstract contract AbstractMessageIdAuthorizedIsm is
+    IInterchainSecurityModule,
+    Initializable
+{
+    using Address for address payable;
+    using LibBit for uint256;
+    using Message for bytes;
+    // ============ Public Storage ============
+
+    /// @notice Maps messageId to whether or not the message has been verified
+    /// first bit is boolean for verification
+    /// rest of bits is the amount to send to the recipient
+    /// @dev bc of the bit packing, we can only send up to 2^255 wei
+    mapping(bytes32 => uint256) public verifiedMessages;
+    /// @notice Index of verification bit in verifiedMessages
+    uint256 public constant MASK_INDEX = 255;
+    /// @notice Address for Hook on L1 responsible for sending message via the Optimism bridge
+    address public authorizedHook;
+
+    // ============ Events ============
+
+    /// @notice Emitted when a message is received from the external bridge
+    /// Might be useful for debugging for the scraper
+    event ReceivedMessage(bytes32 indexed messageId);
+
+    // ============ Initializer ============
+
+    function setAuthorizedHook(address _hook) external initializer {
+        require(
+            _hook != address(0),
+            "AbstractNativeISM: invalid authorized hook"
+        );
+        authorizedHook = _hook;
+    }
+
+    // ============ External Functions ============
+
+    /**
+     * @notice Verify a message was received by ISM.
+     * @param message Message to verify.
+     */
+    function verify(
```

Comment:
I believe this function will need to be `onlyMailbox`

Otherwise Alice can send 1 ETH over the native bridge once, and then whenever Bob sends 1 ETH, in the time between the native bridge calling `verifyMessageId` for Bob's message and when Bob's message is processed via the Mailbox, Alice can call `verify` with her own message ID here and steal Bob's ETH that's being escrowed in this contract

If we make this `onlyMailbox` then we can enforce the variant that this function is only called exactly once for a message. i think we should keep this in mind that any verify function that mutates state will need this

Reply:
@aroralanuk

Reply:
fixed